Prep/src/service/ConfigMergeAndOverrideSolution.java

```diff
@@ -0,0 +1,199 @@
+package service;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import model.MapEntry;
+import util.Printer;
+
+/**
+ * Given two configs, appConfig and defaultConfig (in JSON format),
+ * return a result config containing the merged key value pairs from both the configs.
+ * If there is a conflict, override with appConfig's entries.
+ *
+ * E.g.
+ * appConfig:
+ * {
+ *     key1: val1
+ *     key2: val2
+ *     key4: {
+ *         key41: val41
+ *         key42: val42
+ *     }
+ * }
+ *
+ * defaultConfig:
+ * {
+ *     key2: val2'
+ *     key3: val3'
+ *     key4: {
+ *         key42: val42'
+ *         key43: val43'
+ *     }
+ * }
+ *
+ * resultConfig:
+ * {
+ *     key1: val1
+ *     key2: val2
+ *     key3: val3'
+ *     key4: {
+ *         key41: val41
+ *         key42: val42
+ *         key43: val43'
+ *     }
+ * }
+ *
+ * Stretch Goal:
+ * Flatten the resultConfig:
+ * {
+ *     key1: val1
+ *     key2: val2
+ *     key3: val3'
+ *     key4.key41: val41
+ *     :: ::
+ * }
+ */
+public class ConfigMergeAndOverrideSolution {
+
+    // I am assuming that the JSON input has already been deserialized to Map<String, Object>
+    public static void invoke() {
+        Map<String, Object> appConfig = new HashMap<String, Object>(){{
+            put("key1", "val1");
+            put("key2", "val2");
+            put("key4", new HashMap<String, Object>(){{
+                put("key41", "val41");
+                put("key42", "val42");
+            }});
+        }};
+        Map<String, Object> defaultConfig = new HashMap<String, Object>(){{
+            put("key2", "val2_");
+            put("key3", "val3_");
+            put("key4", new HashMap<String, Object>(){{
+                put("key42", "val42_");
+                put("key43", "val43_");
+            }});
+        }};
+
+        // Prints {key1=val1, key2=val2, key3=val3_, key4={key43=val43_, key42=val42, key41=val41}}
+        Printer.print(configMapMergeAndOverride(appConfig, defaultConfig));
+
+        Printer.println();
+
+        // Prints {key1=val1, key2=val2, key4.key42=val42, key4.key43=val43_, key4.key41=val41, key3=val3_}
+        Printer.print(configMapMergeAndOverrideStretchGoal(appConfig, defaultConfig));
+    }
+
+    // Time complexity: O(n + m) = O(n)
+    // Space Complexity: O(n + m) = O(n)
+    public static Map<String, Object>  configMapMergeAndOverride(
+            Map<String, Object> appConfig,
+            Map<String, Object> defaultConfig) {
+
+        // input validation
+        if (appConfig == null) {
+            return defaultConfig;
+        } else if (defaultConfig == null) {
+            return appConfig;
+        } else {
```

Comment:
Need to also accommodate the case of object being empty.

Reply:
Good call! Added.